### ISSUE TITLE
Configuring PromQL query statistics

### ISSUE THREAD

**Issue by puetzp** [prometheus/prometheus#10953]
Hey there,
it is currently possible to retrieve query statistics  using the `stats` query parameter. Since v2.35 query statistics of range queries can be extended with per-step data by running Prometheus with `--enable-feature=promql-per-step-stats`. 
When I'm not interested in per-step stats I simply need to pass `stats` with in arbitrary non-empty value, e.g. `stats=foo`. So the query parameters basically behaves like a boolean, see [source](https://github.com/prometheus/prometheus/blob/90583c890621ee8c5598adca37c5b9308e67cd6c/web/api/v1/api.go#L109):
```go
func defaultStatsRenderer(ctx context.Context, s *stats.Statistics, param string) stats.QueryStats {
	if param != "" {
		return stats.NewQueryStats(s)
	}
	return nil
}
```

In the context of per-step stats however I need to pass `stats=all` in order to retrieve the data. So in the latter case the value matters and is pattern-matched in the [handler](https://github.com/prometheus/prometheus/blob/13bd4fd3c81c8a483575a183139d830da8d1563e/web/api/v1/api.go#L432): 
```go
func extractQueryOpts(r *http.Request) *promql.QueryOpts {
	return &promql.QueryOpts{
		EnablePerStepStats: r.FormValue("stats") == "all",
	}
}
```
I guess my question is if it does make sense to change `stats` to an enum and deny unsupported values since other endpoints already do this (e.g. `/api/v1/rules` with `type=alert|record`).
I think that would make it easier for client libraries to reason about this parameter as well.

Feel free to close this issue when the matter has been discussed ad nauseam.

**Comment by roidelapluie** [comment]
Hello,

Unfortunately this is not possible within Prometheus v2. We could consider it for Prometheus v3.

**Comment by puetzp** [comment]
That would be great! By the way is there a place where ideas/discussions/issues are tracked for a future v3?

**Comment by beorn7** [comment]
Hello from the bug scrub.

We missed to mark this issue as relevant for v3, sorry for that. (Now it's probably too late…)

However, this might not be seen as a breaking change and could be tweaked anyway. Suggestions welcome.

**Comment by beorn7** [comment]
> V3 is beta, we can make the change.

Yes, if somebody is up to implementing this very soon. The question is more if we want to block the final v3 release on this issue here.

**Comment by flying-musk** [comment]
Hi, I’m interested in working on this issue.

I noticed this issue is old, so before I start, I just wanted to confirm whether this is still something the maintainers would be open to. If so, I’d be happy to work on it. Could this issue be assigned to me?

**Comment by junkerderprovinz** [comment]
I'd like to pick this up. @flying-musk — you asked about this back in May; are you still on it? Happy to step back if so.

Otherwise, the shape I'd propose (following beorn7's "suggestions welcome"): validate `stats` on `/api/v1/query` and `/api/v1/query_range` against the values that actually mean something today — empty (no statistics), `true` (basic statistics, what the built-in UI sends), `all` (basic plus per-step) — and reject anything else with a 400 `bad_data`, matching how `/api/v1/rules` validates `type`. Embedders that install a custom `StatsRenderer` keep accepting their own vocabulary, so downstream projects defining custom values are unaffected. Docs and the OpenAPI schema get the enum, plus a CHANGELOG `[CHANGE]` entry since previously any non-empty value was silently accepted as "basic stats".

I'll open a PR with that shape shortly; happy to adjust the accepted set there if anyone prefers something more lenient.

**Comment by bboreham** [comment]
Thanks for your interest @junkerderprovinz and @flying-musk.
As previously, this change will be breaking for users of Prometheus so could only be done when we make a major version.
I'll add this to the v4 board.
